File: pubnub-gson/pubnub-gson-impl/src/main/java/com/pubnub/internal/java/datasync/DataSyncImpl.java
package com.pubnub.internal.java.datasync;

import com.pubnub.api.PubNub;
import com.pubnub.api.java.datasync.DataSync;
import com.pubnub.api.java.endpoints.datasync.channel.CreateChannel;
import com.pubnub.api.java.endpoints.datasync.channel.GetChannel;
import com.pubnub.api.java.endpoints.datasync.channel.GetChannels;
import com.pubnub.api.java.endpoints.datasync.channel.RemoveChannel;
import com.pubnub.api.java.endpoints.datasync.channel.SetChannel;
import com.pubnub.api.java.endpoints.datasync.channel.UpdateChannel;
import com.pubnub.api.java.endpoints.datasync.entity.CreateEntity;
import com.pubnub.api.java.endpoints.datasync.entity.GetEntities;
import com.pubnub.api.java.endpoints.datasync.entity.GetEntity;
import com.pubnub.api.java.endpoints.datasync.entity.UpdateEntity;
import com.pubnub.api.java.endpoints.datasync.entity.RemoveEntity;
import com.pubnub.api.java.endpoints.datasync.entity.SetEntity;
import com.pubnub.api.java.endpoints.datasync.user.CreateUser;
import com.pubnub.api.java.endpoints.datasync.user.GetUser;
import com.pubnub.api.java.endpoints.datasync.user.GetUsers;
import com.pubnub.api.java.endpoints.datasync.user.UpdateUser;
import com.pubnub.api.java.endpoints.datasync.user.RemoveUser;
import com.pubnub.api.java.endpoints.datasync.user.SetUser;
import com.pubnub.api.java.models.consumer.datasync.entity.PNJsonPatchOperation;
import com.pubnub.internal.java.endpoints.datasync.channel.CreateChannelImpl;
import com.pubnub.internal.java.endpoints.datasync.channel.GetChannelImpl;
import com.pubnub.internal.java.endpoints.datasync.channel.GetChannelsImpl;
import com.pubnub.internal.java.endpoints.datasync.channel.RemoveChannelImpl;
import com.pubnub.internal.java.endpoints.datasync.channel.SetChannelImpl;
import com.pubnub.internal.java.endpoints.datasync.channel.UpdateChannelImpl;
import com.pubnub.internal.java.endpoints.datasync.entity.CreateEntityImpl;
import com.pubnub.internal.java.endpoints.datasync.entity.GetEntitiesImpl;
import com.pubnub.internal.java.endpoints.datasync.entity.GetEntityImpl;
import com.pubnub.internal.java.endpoints.datasync.entity.UpdateEntityImpl;
import com.pubnub.internal.java.endpoints.datasync.entity.RemoveEntityImpl;
import com.pubnub.internal.java.endpoints.datasync.entity.SetEntityImpl;
import com.pubnub.internal.java.endpoints.datasync.user.CreateUserImpl;
import com.pubnub.internal.java.endpoints.datasync.user.GetUserImpl;
import com.pubnub.internal.java.endpoints.datasync.user.GetUsersImpl;
import com.pubnub.internal.java.endpoints.datasync.user.UpdateUserImpl;
import com.pubnub.internal.java.endpoints.datasync.user.RemoveUserImpl;
import com.pubnub.internal.java.endpoints.datasync.user.SetUserImpl;

import java.util.List;

public class DataSyncImpl implements DataSync {
    private final PubNub pubnubInstance;

    public DataSyncImpl(PubNub pubnubInstance) {
        this.pubnubInstance = pubnubInstance;
    }

    @Override
    public GetEntity getEntity(String entityId) {
        return new GetEntityImpl(entityId, pubnubInstance);
    }

    @Override
    public CreateEntity createEntity(String className, int classVersion) {
        return new CreateEntityImpl(className, classVersion, pubnubInstance);
    }

    @Override
    public RemoveEntity removeEntity(String entityId) {
        return new RemoveEntityImpl(entityId, pubnubInstance);
    }

    @Override
    public GetEntities getEntities(String className) {
        return new GetEntitiesImpl(className, pubnubInstance);
    }

    @Override
    public UpdateEntity updateEntity(String entityId, List<PNJsonPatchOperation> operations) {
        return new UpdateEntityImpl(entityId, operations, pubnubInstance);
    }

    @Override
    public SetEntity setEntity(String entityId, int classVersion) {
        return new SetEntityImpl(entityId, classVersion, pubnubInstance);
    }

    @Override
    public GetUser getUser(String userId) {
        return new GetUserImpl(userId, pubnubInstance);
    }

    @Override
    public CreateUser createUser(int classVersion) {
        return new CreateUserImpl(classVersion, pubnubInstance);
    }

    @Override
    public RemoveUser removeUser(String userId) {
        return new RemoveUserImpl(userId, pubnubInstance);
    }

    @Override
    public GetUsers getUsers() {
        return new GetUsersImpl(pubnubInstance);
    }

    @Override
    public UpdateUser updateUser(String userId, List<PNJsonPatchOperation> operations) {
        return new UpdateUserImpl(userId, operations, pubnubInstance);
    }

    @Override
    public SetUser setUser(String userId, int classVersion) {
        return new SetUserImpl(userId, classVersion, pubnubInstance);
    }

    @Override
    public GetChannel getChannel(String channelId) {
        return new GetChannelImpl(channelId, pubnubInstance);
    }

    @Override
    public CreateChannel createChannel(int classVersion) {
        return new CreateChannelImpl(classVersion, pubnubInstance);
    }

    @Override
    public RemoveChannel removeChannel(String channelId) {
        return new RemoveChannelImpl(channelId, pubnubInstance);
    }

    @Override
    public GetChannels getChannels() {
        return new GetChannelsImpl(pubnubInstance);
    }

    @Override
    public UpdateChannel updateChannel(String channelId, List<PNJsonPatchOperation> operations) {
        return new UpdateChannelImpl(channelId, operations, pubnubInstance);
    }

    @Override
    public SetChannel setChannel(String channelId, int classVersion) {
        return new SetChannelImpl(channelId, classVersion, pubnubInstance);
    }
}

File: pubnub-kotlin/pubnub-kotlin-impl/src/main/kotlin/com/pubnub/internal/datasync/DataSyncImpl.kt
package com.pubnub.internal.datasync

import com.pubnub.api.datasync.DataSync
import com.pubnub.api.endpoints.datasync.channel.CreateChannel
import com.pubnub.api.endpoints.datasync.channel.GetChannel
import com.pubnub.api.endpoints.datasync.channel.GetChannels
import com.pubnub.api.endpoints.datasync.channel.RemoveChannel
import com.pubnub.api.endpoints.datasync.channel.SetChannel
import com.pubnub.api.endpoints.datasync.channel.UpdateChannel
import com.pubnub.api.endpoints.datasync.entity.CreateEntity
import com.pubnub.api.endpoints.datasync.entity.GetEntities
import com.pubnub.api.endpoints.datasync.entity.GetEntity
import com.pubnub.api.endpoints.datasync.entity.RemoveEntity
import com.pubnub.api.endpoints.datasync.entity.SetEntity
import com.pubnub.api.endpoints.datasync.entity.UpdateEntity
import com.pubnub.api.endpoints.datasync.membership.CreateMembership
import com.pubnub.api.endpoints.datasync.membership.GetMembership
import com.pubnub.api.endpoints.datasync.membership.GetMemberships
import com.pubnub.api.endpoints.datasync.membership.RemoveMembership
import com.pubnub.api.endpoints.datasync.membership.SetMembership
import com.pubnub.api.endpoints.datasync.membership.UpdateMembership
import com.pubnub.api.endpoints.datasync.user.CreateUser
import com.pubnub.api.endpoints.datasync.user.GetUser
import com.pubnub.api.endpoints.datasync.user.GetUsers
import com.pubnub.api.endpoints.datasync.user.RemoveUser
import com.pubnub.api.endpoints.datasync.user.SetUser
import com.pubnub.api.endpoints.datasync.user.UpdateUser
import com.pubnub.api.models.consumer.datasync.PNDataSyncClassLevel
import com.pubnub.api.models.consumer.datasync.PNDataSyncSortField
import com.pubnub.api.models.consumer.datasync.entity.PNJsonPatchOperation
import com.pubnub.internal.PubNubImpl
import com.pubnub.internal.endpoints.datasync.channel.CreateChannelEndpoint
import com.pubnub.internal.endpoints.datasync.channel.GetChannelEndpoint
import com.pubnub.internal.endpoints.datasync.channel.GetChannelsEndpoint
import com.pubnub.internal.endpoints.datasync.channel.RemoveChannelEndpoint
import com.pubnub.internal.endpoints.datasync.channel.SetChannelEndpoint
import com.pubnub.internal.endpoints.datasync.channel.UpdateChannelEndpoint
import com.pubnub.internal.endpoints.datasync.entity.CreateEntityEndpoint
import com.pubnub.internal.endpoints.datasync.entity.GetEntitiesEndpoint
import com.pubnub.internal.endpoints.datasync.entity.GetEntityEndpoint
import com.pubnub.internal.endpoints.datasync.entity.RemoveEntityEndpoint
import com.pubnub.internal.endpoints.datasync.entity.SetEntityEndpoint
import com.pubnub.internal.endpoints.datasync.entity.UpdateEntityEndpoint
import com.pubnub.internal.endpoints.datasync.membership.CreateMembershipEndpoint
import com.pubnub.internal.endpoints.datasync.membership.GetMembershipEndpoint
import com.pubnub.internal.endpoints.datasync.membership.GetMembershipsEndpoint
import com.pubnub.internal.endpoints.datasync.membership.RemoveMembershipEndpoint
import com.pubnub.internal.endpoints.datasync.membership.SetMembershipEndpoint
import com.pubnub.internal.endpoints.datasync.membership.UpdateMembershipEndpoint
import com.pubnub.internal.endpoints.datasync.user.CreateUserEndpoint
import com.pubnub.internal.endpoints.datasync.user.GetUserEndpoint
import com.pubnub.internal.endpoints.datasync.user.GetUsersEndpoint
import com.pubnub.internal.endpoints.datasync.user.RemoveUserEndpoint
import com.pubnub.internal.endpoints.datasync.user.SetUserEndpoint
import com.pubnub.internal.endpoints.datasync.user.UpdateUserEndpoint

class DataSyncImpl(private val pubnub: PubNubImpl) : DataSync {
    override fun getEntity(entityId: String): GetEntity {
        return GetEntityEndpoint(pubnub, entityId)
    }

    override fun createEntity(
        className: String,
        classVersion: Int,
        classLevel: PNDataSyncClassLevel?,
        entityId: String?,
        status: String?,
        payload: Any?,
    ): CreateEntity {
        return CreateEntityEndpoint(
            pubnub = pubnub,
            className = className,
            classVersion = classVersion,
            classLevel = classLevel,
            entityId = entityId,
            status = status,
            payload = payload,
        )
    }

    override fun removeEntity(entityId: String, ifMatch: String?): RemoveEntity {
        return RemoveEntityEndpoint(pubnub, entityId, ifMatch)
    }

    override fun getEntities(
        className: String,
        classVersion: Int?,
        classLevel: PNDataSyncClassLevel?,
        filterFast: String?,
        filter: String?,
        sort: List<PNDataSyncSortField>,
        limit: Int?,
        cursor: String?,
    ): GetEntities {
        return GetEntitiesEndpoint(
            pubnub = pubnub,
            className = className,
            classVersion = classVersion,
            classLevel = classLevel,
            filterFast = filterFast,
            filter = filter,
            sort = sort,
            limit = limit,
            cursor = cursor,
        )
    }

    override fun updateEntity(
        entityId: String,
        operations: List<PNJsonPatchOperation>,
        ifMatch: String?,
    ): UpdateEntity {
        return UpdateEntityEndpoint(pubnub, entityId, operations, ifMatch)
    }

    override fun setEntity(
        entityId: String,
        classVersion: Int,
        status: String?,
        payload: Any?,
        ifMatch: String?,
    ): SetEntity {
        return SetEntityEndpoint(
            pubnub = pubnub,
            entityId = entityId,
            classVersion = classVersion,
            status = status,
            payload = payload,
            ifMatch = ifMatch,
        )
    }

    override fun getUser(userId: String): GetUser {
        return GetUserEndpoint(pubnub, userId)
    }

    override fun createUser(
        classVersion: Int,
        userId: String?,
        className: String?,
        classLevel: PNDataSyncClassLevel?,
        status: String?,
        payload: Any?,
    ): CreateUser {
        return CreateUserEndpoint(
            pubnub = pubnub,
            className = className,
            classVersion = classVersion,
            classLevel = classLevel,
            userId = userId,
            status = status,
            payload = payload,
        )
    }

    override fun removeUser(userId: String, ifMatch: String?): RemoveUser {
        return RemoveUserEndpoint(pubnub, userId, ifMatch)
    }

    override fun getUsers(
        className: String?,
        classVersion: Int?,
        classLevel: PNDataSyncClassLevel?,
        filterFast: String?,
        filter: String?,
        sort: List<PNDataSyncSortField>,
        limit: Int?,
        cursor: String?,
    ): GetUsers {
        return GetUsersEndpoint(
            pubnub = pubnub,
            className = className,
            classVersion = classVersion,
            classLevel = classLevel,
            filterFast = filterFast,
            filter = filter,
            sort = sort,
            limit = limit,
            cursor = cursor,
        )
    }

    override fun updateUser(
        userId: String,
        operations: List<PNJsonPatchOperation>,
        ifMatch: String?,
    ): UpdateUser {
        return UpdateUserEndpoint(pubnub, userId, operations, ifMatch)
    }

    override fun setUser(
        userId: String,
        classVersion: Int,
        status: String?,
        payload: Any?,
        ifMatch: String?,
    ): SetUser {
        return SetUserEndpoint(
            pubnub = pubnub,
            userId = userId,
            classVersion = classVersion,
            status = status,
            payload = payload,
            ifMatch = ifMatch,
        )
    }

    override fun getChannel(channelId: String): GetChannel {
        return GetChannelEndpoint(pubnub, channelId)
    }

    override fun createChannel(
        classVersion: Int,
        channelId: String?,
        className: String?,
        classLevel: PNDataSyncClassLevel?,
        status: String?,
        payload: Any?,
    ): CreateChannel {
        return CreateChannelEndpoint(
            pubnub = pubnub,
            className = className,
            classVersion = classVersion,
            classLevel = classLevel,
            channelId = channelId,
            status = status,
            payload = payload,
        )
    }

    override fun removeChannel(channelId: String, ifMatch: String?): RemoveChannel {
        return RemoveChannelEndpoint(pubnub, channelId, ifMatch)
    }

    override fun getChannels(
        className: String?,
        classVersion: Int?,
        classLevel: PNDataSyncClassLevel?,
        filterFast: String?,
        filter: String?,
        sort: List<PNDataSyncSortField>,
        limit: Int?,
        cursor: String?,
    ): GetChannels {
        return GetChannelsEndpoint(
            pubnub = pubnub,
            className = className,
            classVersion = classVersion,
            classLevel = classLevel,
            filterFast = filterFast,
            filter = filter,
            sort = sort,
            limit = limit,
            cursor = cursor,
        )
    }

    override fun updateChannel(
        channelId: String,
        operations: List<PNJsonPatchOperation>,
        ifMatch: String?,
    ): UpdateChannel {
        return UpdateChannelEndpoint(pubnub, channelId, operations, ifMatch)
    }

    override fun setChannel(
        channelId: String,
        classVersion: Int,
        status: String?,
        payload: Any?,
        ifMatch: String?,
    ): SetChannel {
        return SetChannelEndpoint(
            pubnub = pubnub,
            channelId = channelId,
            classVersion = classVersion,
            status = status,
            payload = payload,
            ifMatch = ifMatch,
        )
    }

    override fun getMembership(membershipId: String): GetMembership {
        return GetMembershipEndpoint(pubnub, membershipId)
    }

    override fun createMembership(
        channelId: String,
        userId: String,
        classVersion: Int,
        membershipId: String?,
        status: String?,
        payload: Any?,
    ): CreateMembership {
        return CreateMembershipEndpoint(
            pubnub = pubnub,
            channelId = channelId,
            userId = userId,
            classVersion = classVersion,
            membershipId = membershipId,
            status = status,
            payload = payload,
        )
    }

    override fun removeMembership(membershipId: String, ifMatch: String?): RemoveMembership {
        return RemoveMembershipEndpoint(pubnub, membershipId, ifMatch)
    }

    override fun getMemberships(
        channelId: String?,
        userId: String?,
        classVersion: Int?,
        filterFast: String?,
        filter: String?,
        sort: List<PNDataSyncSortField>,
        limit: Int?,
        cursor: String?,
    ): GetMemberships {
        return GetMembershipsEndpoint(
            pubnub = pubnub,
            channelId = channelId,
            userId = userId,
            classVersion = classVersion,
            filterFast = filterFast,
            filter = filter,
            sort = sort,
            limit = limit,
            cursor = cursor,
        )
    }

    override fun updateMembership(
        membershipId: String,
        operations: List<PNJsonPatchOperation>,
        ifMatch: String?,
    ): UpdateMembership {
        return UpdateMembershipEndpoint(pubnub, membershipId, operations, ifMatch)
    }

    override fun setMembership(
        membershipId: String,
        classVersion: Int,
        status: String?,
        payload: Any?,
        ifMatch: String?,
    ): SetMembership {
        return SetMembershipEndpoint(
            pubnub = pubnub,
            membershipId = membershipId,
            classVersion = classVersion,
            status = status,
            payload = payload,
            ifMatch = ifMatch,
        )
    }
}
